File: 8/py/treetop_calculator.py
def run(path='../map.txt'):
    file_data = open(path).read().strip().split('\n')
    grid = [list(map(int, line)) for line in file_data]

    visible_trees, tree_scores = treetop_tree_house(grid)
    print(visible_trees)
    print(tree_scores)

def treetop_tree_house(grid):
    total, tree_values = (4*len(grid)) - 4, []
    for i in range(1, len(grid)-1):
        for j in range(1, len(grid[i])-1):
            item = grid[i][j]
            up = [grid[k][j] for k in range(0, i)]
            down = [grid[k][j] for k in range(i+1, len(grid[i]))]
            left = grid[i][0:j]
            right = grid[i][j+1:len(grid[j])]
            if check_visible(item, [up[::-1], down, left[::-1], right]): total += 1
            tree_values.append(view_distance(item, up[::-1]) * view_distance(item, down) * view_distance(item, left[::-1]) * view_distance(item, right))

    return total, max(tree_values)

def check_visible(item, directions):
    for direction in directions:
        row_val = 0
        for i in range(0, len(direction)):
            if item <= direction[i]: row_val = 1
        if row_val == 0: return True
    return False

def view_distance(item, direction):
    v_distance = 0
    for i in range(0, len(direction)):
        v_distance += 1
        if item <= direction[i]: break
    return v_distance

File: 8/rb/treetop_calculator.rb
def run(path = "../map.txt")
    data = File.read(path).split("\n")
    grid = make_grid(data)
    
    puts tree_house(grid)
end

def make_grid(data, grid = [])
    data.each_with_index do |line, i|
        grid.append([])
        chars = line.chars
        chars.each do |num|
            grid[i].append(num.to_i)
        end
    end
    return grid
end

def tree_house(grid)
    total, tree_values = (4 * grid.length) - 4, []
    for i in 1..grid.length-2 do
        for j in 1..grid[i].length-2 do
            item = grid[i][j]

            up, down, left, right = [], [], grid[i][0..j-1], grid[i][j+1..grid[j].length]
            for k in 0..i-1 do up.append(grid[k][j]) end
            for k in i+1..grid[i].length-1 do down.append(grid[k][j]) end
            
            total += 1 if check_visible?(item, [up, down, left, right])
            tree_values.append(view_distance(item, up.reverse()) * view_distance(item, down) * view_distance(item, left.reverse()) * view_distance(item, right))
        end
    end
    return total, tree_values.max()
end

def check_visible?(item, directions)
    directions.each do |direction|
        blocked = false
        direction.each do |tree|
            blocked = true if item <= tree
        end
        return true if not blocked
    end
    return false
end

def view_distance(item, direction)
    v_distance = 0
    direction.each do |dir|
        v_distance += 1
        break if item <= dir
    end
    return v_distance
end